sourcegraph/sourcegraph
# Cody: Add support for connecting Cody extension to App

Adds support for connecting the Cody VSCode extension to Sourcegraph App.

### "Connect Sourcegraph App" button

<img width="402" alt="Screenshot 2023-05-11 at 11 53 20 AM" src="https://github.com/sourcegraph/sourcegraph/assets/602886/19e72e88-e90d-4751-84fb-68d7d691f83c">

The "Connect Sourcegraph App" feature is behind an experimental flag, `cody.experimental.connectToApp`.

The feature is enabled if App is detected as being installed on the system. This currently works on mac OS. If App is not detected as installed, the button shows "Get Sourcegraph App" instead.

### Local app detector

The `LocalAppDetector` detects whether App is installed. It does this by checking for the existence of known directory locations in the file system.

It uses polling, and for that reason it can be controlled with `start` and `stop` methods. It will stop detecting if the user is authed, or if an App installation has been detected.

### Connection flow

To connect to App, the extension sends the user to App with a sourcegraph:// url to generate a token.

App then sends the user back to the extension using the token callback mechanism.

After successfully connecting:
- the `serverEndpoint` config is set to the local App's endpoint which is `http://localhost:3080`
- the `access-token` secret is set to the access token obtained from App.

## Test plan

- Run a build of Sourcegraph App (it won't work with the dev `sg start app` version)
- Click on "Connect Sourcegraph App" in the extension
- Verify that the flow is working
- Verify that the extension now has a token that works to perform request to the local App instance.
